### PR TITLE
Health check for regions

### PR DESCRIPTION
**Pull request by huachaohuang**
We will do balance for region if:

- the peer is down for too long (`>= MaxPeerDownDuration`); or
- the peer is down for a while and its store is down too.

### CODE REVIEW

File: server/balancer.go [hunk 1]
```diff
@@ -417,3 +418,75 @@ func (db *defaultBalancer) Balance(cluster *clusterInfo) (*score, *balanceOperat
...
+	for _, stats := range rb.downPeers {
+		peer := stats.GetPeer()
+		if peer == nil {
+			continue
+		}
+		store := cluster.getStore(peer.GetStoreId())
+		if store == nil {
+			continue
+		}
+		if stats.GetDownSeconds() >= rb.cfg.MaxPeerDownDuration {
```
Comment (siddontang) [inline server/balancer.go]: Each time I see Duration, I think this is time.Duration, but here is seconds.
maybe we should think a better name? @qiuyesuifeng

Comment (qiuyesuifeng) [inline server/balancer.go]: I prefer MaxPeerDownTime, it can be defined as a time.Second.

Comment (huachaohuang) [inline server/balancer.go]: I think I should use `time.Duration` for the configurations.

File: server/balancer.go [hunk 2]
```diff
@@ -417,3 +418,75 @@ func (db *defaultBalancer) Balance(cluster *clusterInfo) (*score, *balanceOperat
...
 	return nil, bop, errors.Trace(err)
 }
+
+// replicaBalancer is used to balance active replica count.
+type replicaBalancer struct {
+	cfg             *BalanceConfig
+	region          *metapb.Region
+	leader          *metapb.Peer
+	downPeers       []*pdpb.PeerStats
+	defaultBalancer *defaultBalancer
```
Comment (qiuyesuifeng) [inline server/balancer.go]: Seems defaultBalancer can be removed by using replicaBalancer instead?

Comment (huachaohuang) [inline server/balancer.go]: I think we can keep it and test them separately?

File: server/config.go [hunk 3]
```diff
@@ -345,8 +364,12 @@ func (c *BalanceConfig) adjust() {
 		c.MaxTransferWaitCount = defaultMaxTransferWaitCount
 	}
 
-	if c.MaxStoreDownDuration == 0 {
-		c.MaxStoreDownDuration = defaultMaxStoreDownDuration
+	if c.MaxPeerDownDuration.Duration == 0 {
```
Comment (siddontang) [inline server/config.go]: extract to a adjustDuration function, if 0, use default

File: server/config.go [hunk 4]
```diff
@@ -297,7 +315,8 @@ const (
 	defaultMaxBalanceRetryPerLoop = uint64(10)
 	defaultMaxBalanceCountPerLoop = uint64(3)
 	defaultMaxTransferWaitCount   = uint64(3)
-	defaultMaxStoreDownDuration   = uint64(60)
+	defaultMaxPeerDownDuration    = 30 * time.Minute
+	defaultMaxStoreDownDuration   = 60 * time.Second
```
Comment (siddontang) [inline server/config.go]: the default peer down is 30m, but store down is 60s, too small?

File: server/balancer.go [hunk 5]
```diff
@@ -417,3 +418,75 @@ func (db *defaultBalancer) Balance(cluster *clusterInfo) (*score, *balanceOperat
...
+func (rb *replicaBalancer) removePeer(cluster *clusterInfo, downPeers []*metapb.Peer) (*balanceOperator, error) {
+	if len(downPeers) >= 1 {
+		removePeerOperator := newRemovePeerOperator(rb.region.GetId(), downPeers[0])
+		return newBalanceOperator(rb.region, newOnceOperator(removePeerOperator)), nil
+	}
+	return rb.defaultBalancer.removePeer(cluster)
+}
+
+func (rb *replicaBalancer) collectDownPeers(cluster *clusterInfo) []*metapb.Peer {
+	var downPeers []*metapb.Peer
```
Comment (qiuyesuifeng) [inline server/balancer.go]: I suggest we use `downPeers  := make([]*metapb.Peer,0,len(rb.downPeers))`.

File: server/balancer.go (hunk 2 above)
Comment (qiuyesuifeng) [inline server/balancer.go]: But seems that defaultBalancer's Balance will not be called since this PR?

File: server/config.go [hunk 6]
```diff
@@ -234,6 +246,20 @@ func (c *Config) configFromFile(path string) error {
...
+}
+
+func (d *duration) Seconds() uint64 {
+	return uint64(d.Duration.Seconds())
+}
+
+func (d *duration) UnmarshalText(text []byte) error {
+	var err error
+	d.Duration, err = time.ParseDuration(string(text))
+	return err
```
Comment (qiuyesuifeng) [inline server/config.go]: errors.Trace(err)

File: conf/config.toml [hunk 7]
```diff
@@ -40,4 +40,5 @@ max-balance-count = 16
 max-balance-retry-per-loop = 10
 max-balance-count-per-loop = 3
 max-transfer-wait-count = 3
-max-store-down-duration = 60
+max-peer-down-duration = "30m"
```
Comment (qiuyesuifeng) [inline conf/config.toml]: I think we should discuss the max-peer-down-duration variable name, it is confused.

Comment (siddontang) [inline conf/config.toml]: I think both time and duration are ok.

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**